File: 384-shuffle-an-array/shuffle-an-array.py
# choose any random number, swap it with last element in array and pop the last element
import random

class Solution:
    def __init__(self, nums):
        self.nums = nums[:]

    def reset(self):
        return self.nums[:]

    def shuffle(self):
        temp = self.nums[:]
        result = []

        while temp:
            r = random.randrange(len(temp))
            result.append(temp[r])
            temp[r], temp[-1] = temp[-1], temp[r]
            temp.pop()
        return result

File: 384-shuffle-an-array/shuffle-an-array.cpp
class Solution {
public:
    vector<int>a;
    vector<int>officialNums;

    Solution(vector<int>& nums) {
        a = nums;
        officialNums = nums;
    }
    
    vector<int> reset() {
        return officialNums;
    }
    
    vector<int> shuffle() {
        a = officialNums;
        vector<int>ss;
        while(a.size()){
            int r = rand()%a.size();
            ss.push_back(a[r]);
            swap(a[r], a[a.size()-1]);
            a.pop_back();
        }
        return ss;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(nums);
 * vector<int> param_1 = obj->reset();
 * vector<int> param_2 = obj->shuffle();
 */